### DISCUSSION TITLE
A question about elements_from_sequence in parttree

### DISCUSSION THREAD

**Discussion by mayy17** [mjhoptics/ray-optics#114] [Q&A] [answered]
I see 

> node = part_tree.parent_node((seq_model.gaps[-1], seq_model.z_dir[-1]))

 in elements_from_sequence() in parttree
But it seems that z_dir refers to dir of mirrors instead of gaps, so len(z_dir)=len(gaps)+1.
In my optic system with five mirrors, the z_dir of Img surf is +1(maybe in default?), but z_dir[-2]=-1, and actually the z_dir of airgap is -1 . It causes the tree cannot find the node.
Maybe it should be

> node = part_tree.parent_node((seq_model.gaps[-1], seq_model.z_dir[-2]))
?

**Comment by mjhoptics** [comment]
Hello @mayy17,

Thank you for the report. The intent of the code is that gaps, z_dir, and rndx are all the same length, and ifcs, lcl_tfrms, and gbl_tfrms are the same and longer by 1. Having said that, it hasn't always been that way and I can imagine there are some code paths that produce lists that don't conform to this intent.

I tried several models but wasn't able to duplicate the problem. It would be helpful if:

1) clarify which version of ray-optics you're using
2) you could post the model (or a dumbed down version if sensitive) and/or script that produces the error
3) If  item 2 isn't possible, please clarify how the model was generated, i.e. Zemax or CODE V file, .roa file, direct model construction via script (using either add_surface or the opm methods), or ...

Thank you for using ray-optics.
Regards,
Mike Hayford

**Reply by mayy17** [reply]
So thanks for your reply and your works.
I create my model in zemax 19.4 and try to read it with rayoptics==0.8.5. And I also tried code in [https://ray-optics.readthedocs.io/en/latest/_modules/rayoptics/zemax/zmxread.html#read_lens_file](url) but got the same result.
I've attached a simplified model which may help. I succeed to duplicate the problem with it. 
[LENS.zip](https://github.com/mjhoptics/ray-optics/files/11206976/LENS.zip)

And with the previous code I can read the model and get result of raytrace, maybe this problem is just caused by a bug which creates a wrong model in read_lens_file() in zmxread.

**Reply by mjhoptics** [reply, marked as answer]
Hi @mayy17,
I just updated a fix to the master branch; the postprocessing of the zemax file import needed tweaking. The commit is f055ded7e39ccf1fca54a1c1ec2f53f66a1705b0.
HTH
Mike

**Reply by mayy17** [reply]
It works well. Thanks for your help.
